canonicalwebteam/http/__init__.py

```diff
@@ -0,0 +1,105 @@
+# Core packages
+import os
+from urllib.parse import urlparse
+
+# Third-party packages
+import requests
+import requests_cache
+
+try:
+    # If prometheus is available, set up metric counters
+
+    import prometheus_client
+    timeout_counter = prometheus_client.Counter(
+        'feed_timeouts',
+        'A counter of timed out requests',
+        ['domain'],
+    )
+    connection_failed_counter = prometheus_client.Counter(
+        'feed_connection_failures',
+        'A counter of requests which failed to connect',
+        ['domain'],
+    )
+    latency_histogram = prometheus_client.Histogram(
+        'feed_latency_seconds',
+        'Feed requests retrieved',
+        ['domain', 'code'],
+        buckets=[0.25, 0.5, 0.75, 1, 2, 3],
+    )
+except ImportError:
+    timeout_counter = None
+    connection_failed_counter = None
+    latency_histogram = None
+
+
+class TimeoutHTTPAdapter(requests.adapters.HTTPAdapter):
+    """
+    A simple extension to the HTTPAdapter to add a 'timeout' parameter
+    """
+
+    def __init__(self, timeout=None, *args, **kwargs):
+        self.timeout = timeout
+        super().__init__(*args, **kwargs)
+
+    def send(self, *args, **kwargs):
+        kwargs['timeout'] = self.timeout
+        return super().send(*args, **kwargs)
+
+
+class BaseSession():
+    """
+    A base session interface to implement common functionality:
```

Comment:
Worth having a look at PEP257 (https://www.python.org/dev/peps/pep-0257/), for docstring conventions. 

Typically a one line docstring will be in the form:

class Foo():
"""This is a single line docstring."""

and a multiline docstring in the form:

class Bar():
"""An initial description.

More details...
"""

Reply:
Yeah I'm aware that that's the more common format but it's really ugly to me - I find it immediately off-putting.

I find the way I've done it much more readable, and since it's not something that any linter I'm aware of is opinionated about and I don't see much of a downside to minor inconsistency in this case, I'd rather just keep it this way unless we as a team decide we need to be super strict about it (which I hope we don't).

Reply:
Not hugely important, as long as you're aware that it is unconventional.

Reply:
I have been following that convention when I write docstrings but it is probably worth discussing about properly outside this PR.